gremlin-server/src/main/java/org/apache/tinkerpop/gremlin/server/handler/HttpHandlerUtil.java

```diff
@@ -74,6 +78,39 @@ public class HttpHandlerUtil {
     public static RequestMessage getRequestMessageFromHttpRequest(final FullHttpRequest request) {
         final String contentType = Optional.ofNullable(request.headers().get(HttpHeaderNames.CONTENT_TYPE)).orElse("application/json");
 
+        if (contentType.equals(SerTokens.MIME_GRAPHSON_V3)) {
+            final GraphSONMessageSerializerV3 serializer = new GraphSONMessageSerializerV3();
+            try {
+                final ByteBuf buffer = request.content();
+
+                final int first = buffer.readByte();
+                // !!! skip header. Why we write this to json?
```

Comment:
like every odd looking design choice, i think it goes back to an old concept. the gremlin server protocol was written such that the request message included the content type in the serialized message itself: 

https://tinkerpop.apache.org/docs/3.7.0/dev/provider/#_graph_driver_provider_requirements

in this way the server could detect the serializer used for the message and choose the right one to decode it.  For HTTP it is a bit redundant to do this because we supply `Content-Type` explicitly, but when we had a pure NIO based protocol to support we used that as the means for supplying this information. It also stuck for websockets where it is still in use, though in retrospect we probably could have passed that information once in the handshake.

Interestingly, the fact that we encode the serializer here offers a possible solution to some issues faced here with this change. The `Content-Type` for the payload on the `POST` could always stay `application/json` and preserve the old functionality because the serialized `RequestMessage` itself contains the serializer to use to decode it. There would be no need to add an custom header to tell the server what serializer to use. Serializing a whole `RequestMessage` into the JSON request structure for HTTP seems like a fairly clean solution doesn't it?